### PR TITLE
Fixed race condition in map iteration and map write in Discovery (#3735)

### PR DESCRIPTION
**Pull request by codesome**
Added lock for this loop 
[https://github.com/prometheus/prometheus/blob/master/discovery/file/file.go#L99-L101](https://github.com/prometheus/prometheus/blob/master/discovery/file/file.go#L99-L101) 
using `fileSD.lock`

(Issue [#3735](https://github.com/prometheus/prometheus/issues/3735))

### CODE REVIEW

File: discovery/file/file.go [hunk 1]
```diff
@@ -96,9 +96,11 @@ func (t *TimestampCollector) Collect(ch chan<- prometheus.Metric) {
 	uniqueFiles := make(map[string]float64)
 	t.lock.RLock()
 	for fileSD := range t.discoverers {
+		fileSD.lock.Lock()
```
Comment (Gouthamve) [inline discovery/file/file.go]: This would cause rapid locking and unlocking. Can you put it outside the loop?

Comment (Gouthamve) [reply discovery/file/file.go]: Sorry, not sure what I was thinking. Looks good :)

**Comment by cstyan** [comment]
hmmm, read/write locks should have worked :|

**Comment by Gouthamve** [comment]
@cstyan https://github.com/prometheus/prometheus/issues/3735#issuecomment-360455165

Yes, but the locks are different :P

**Comment by cstyan** [comment]
ooooooooohhh, oh boy

yeah, good catch! @Gouthamve

**Comment by codesome** [comment]
@cstyan yes, did not notice that it was `RWMutex`. Changed it to `RLock` now.

**Comment by cstyan** [comment]
this looks good to me

**Pull request merged**